### ISSUE TITLE
Conditional registration for closed-generic types replace earlier registrations instead of being appended

### ISSUE THREAD

**Issue by onyxmaster** [simpleinjector/SimpleInjector#186]
The `RegisterConditional` for closed generic service types behaves strangely.
Namely, when creating multiple registrations, only the last one is considered.

The repo at https://github.com/onyxmaster/sitest has some code that demonstrates the error:
```
Non-generic test:
Tested NonGenericConsumerA against NonGenericConsumerA, matched: True, handled: False.
Tested NonGenericConsumerA against NonGenericConsumerB, matched: False, handled: True.
Tested NonGenericConsumerA against NonGenericConsumerA, matched: True, handled: False.
Tested NonGenericConsumerA against NonGenericConsumerB, matched: False, handled: True.
Tested NonGenericConsumerB against NonGenericConsumerA, matched: False, handled: False.
Tested NonGenericConsumerB against NonGenericConsumerB, matched: True, handled: False.
Tested NonGenericConsumerB against NonGenericConsumerA, matched: False, handled: False.
Tested NonGenericConsumerB against NonGenericConsumerB, matched: True, handled: False.
Generic test:
Tested GenericConsumerA against GenericConsumerB, matched: False, handled: False.
The constructor of type GenericConsumerA contains the parameter with name '_' and type
IGenericInterface<Object> that is not registered. Please ensure IGenericInterface<Object> is
registered, or change the constructor of GenericConsumerA. 1 conditional registration for 
IGenericInterface<T> exists that is applicable to IGenericInterface<Object>, but its supplied 
predicate didn't return true when provided with the contextual information for GenericConsumerA.
```

Edit: moved question not pertinent to the issue to #187

**Comment by qujck** [comment]
Please can you elaborate on what you mean by *"only the last one is considered"*?

For info all Conditional options will be evaluated by the container to ensure there is no overlap. I.e. if you register 6 conditions and the 1st of those conditions returns `true` the other 5 conditions will still be evaluated to ensure they all do not return `true`.

**Comment by onyxmaster** [comment]
To simplify: when I register two conditionals for the same service type with different predicates, if service type is a [closed] generic one, Simple Injector fails to check both predicates, which leads to its inability to resolve one of the services, leading to the exception displayed above. When the service type is non-generic, there is no problem.

As for the post scriptum -- I'll create another issue to keep this issue focused on the problem.

**Comment by qujck** [comment]
I can't figure out if this is a bug or not because you are trying to create 2 conditionals for the same registration. So although you have written code that can fail:

```csharp
var container = new Container();
container.RegisterSingleton<GenericConsumerA>();
container.RegisterSingleton<GenericConsumerB>();
var registration = Lifestyle.Singleton.CreateRegistration(
    () => new ImplOfGenericInterface(), 
    container);
container.RegisterConditional(typeof(IGenericInterface<object>), registration,
    context => context.Consumer.ImplementationType == typeof(GenericConsumerA));
container.RegisterConditional(typeof(IGenericInterface<object>), registration,
    context => context.Consumer.ImplementationType == typeof(GenericConsumerB));
container.Verify();
```

The same code can easily be made to work (and is arguably more logical)

```csharp
var container = new Container();
container.RegisterSingleton<GenericConsumerA>();
container.RegisterSingleton<GenericConsumerB>();
var registration = Lifestyle.Singleton.CreateRegistration(
    () => new ImplOfGenericInterface(), 
    container);
container.RegisterConditional(typeof(IGenericInterface<object>), registration,
    context => context.Consumer.ImplementationType == typeof(GenericConsumerA) 
        || context.Consumer.ImplementationType == typeof(GenericConsumerB));
container.Verify();
 ```

What is the real use case you are struggling with?

**Comment by onyxmaster** [comment]
This is a testing code, so yes, the registration looks similar.
Imagine that `() => new ImplOfGenericInterfaceA()` is different for different consumers.
I'm trying to create implementation-dependent registrations, so different implementations of service types (e.g. abstract database contexts) are resolved for different consumers.

Also, even if I want something that is not considered a good design, the difference between generic and non-generic service type handling is still unexplained.

**Comment by qujck** [comment]
But why are you trying to create 2 registrations for the same closed generic

```csharp
container.RegisterConditional(typeof(IGenericInterface<object>), registration,
    context => context.Consumer.ImplementationType == typeof(GenericConsumerA));
container.RegisterConditional(typeof(IGenericInterface<object>), registration,
    context => context.Consumer.ImplementationType == typeof(GenericConsumerB));
```

rather than put all the logic into one registration

```csharp
var registration = Lifestyle.Singleton.CreateRegistration(
    () => new ImplOfGenericInterface(), 
    container);
container.RegisterConditional(typeof(IGenericInterface<object>), registration,
    context => context.Consumer.ImplementationType == typeof(GenericConsumerA) 
        || context.Consumer.ImplementationType == typeof(GenericConsumerB));
```

**Comment by onyxmaster** [comment]
Because the registrations are actually different, as I indicated in the latest comment.
I've updated the example code to better illustrate what I'm trying to achieve.

**Comment by qujck** [comment]
Sure. The following code fails

```csharp
class ImplOfGenericInterfaceA : IGenericInterface<object>
{
}

class ImplOfGenericInterfaceB : IGenericInterface<object>
{
}

class GenericConsumerA
{
    public GenericConsumerA(IGenericInterface<object> _)
    {
    }
}

class GenericConsumerB
{
    public GenericConsumerB(IGenericInterface<object> _)
    {
    }
}

var container = new Container();
container.RegisterConditional<IGenericInterface<object>, ImplOfGenericInterfaceA>(
    c => c.Consumer.ImplementationType == typeof(GenericConsumerA));
container.RegisterConditional<IGenericInterface<object>, ImplOfGenericInterfaceB>(
    c => c.Consumer.ImplementationType == typeof(GenericConsumerB));
container.Verify();
container.GetInstance<GenericConsumerA>();
container.GetInstance<GenericConsumerB>();
```

**Comment by TheBigRic** [comment]
I don't think this is a bug. When working with generic types `RegisterConditional()` should be used with the open generic type definition instead of supplying a closed generic type. Simple Injector is smart enough to match the closed type at runtime and will probably even stick to any generic type constraints that may be present at a certain implementation.

So this test succeeds:

``` c#
[TestMethod]
public void TestMethod1()
{
    var container = new Container();

    container.RegisterConditional(typeof(IGenericInterface<>), typeof(ImplOfGenericInterfaceA),
        c => c.Consumer.ImplementationType == typeof(GenericConsumerA));
    container.RegisterConditional(typeof(IGenericInterface<>), typeof(ImplOfGenericInterfaceB),
        c => c.Consumer.ImplementationType == typeof(GenericConsumerB));

    container.Register<GenericConsumerA>();
    container.Register<GenericConsumerB>();

    container.Verify();
}
```

Although this could be made more clear in the documentation, the documentation is correct in the examples on this: [How to register open generic types](https://simpleinjector.readthedocs.org/en/latest/advanced.html#registration-of-open-generic-types)

**Comment by onyxmaster** [comment]
If this is not an intended usage pattern, I would expect RegisterConditional to throw when a closed generic type is passed, instead of failing to register the conditional itself and throwing later when resolving/verifying.

**Comment by onyxmaster** [comment]
Also, the restriction on providing closed generic types will, in my opinion, complicate automatic registration with automatic type deduction based on factory functions.

**Comment by TheBigRic** [comment]
> If this is not an intended usage pattern, I would expect RegisterConditional to throw when a closed generic type is passed

Absolutely agreed!

> Also, the restriction on providing closed generic types will, in my opinion, complicate registration with automatic type deduction based on factory functions.

I don't fully understand what you mean. Can you provide an example supporting this comment?

**Comment by dotnetjunkie** [comment]
My impression is that this is s bug.

**Comment by onyxmaster** [comment]
The following example might be too complex to illustrate, but it's what I have now:
```csharp
    public sealed class FactoryRegistration<TTarget>
    {
        private readonly Container _container;

        private readonly Lifestyle _lifestyle;

        private static readonly Predicate<PredicateContext> TargetTypePredicate = GetTargetTypePredicate();

        private static Predicate<PredicateContext> GetTargetTypePredicate()
        {
            var targetType = typeof(TTarget);
            if (targetType.IsInterface || targetType.IsAbstract)
                return context => context.Consumer.ServiceType == targetType;
            return context => context.Consumer.ImplementationType == targetType;
        }

        public void Register<TService>(Func<Container, TService> creator)
            where TService : class
        {
            var registration = _lifestyle.CreateRegistration(
                () => creator(_container),
                _container);
            _container.RegisterConditional(typeof(TService), registration, TargetTypePredicate);
        }

        internal FactoryRegistration(Container container, Lifestyle lifestyle)
        {
            _container = container;
            _lifestyle = lifestyle;
        }
    }

    public static class FactoryRegistrationExtensions
    {
        public static FactoryRegistration<TTarget> FactoryFor<TTarget>(this Container container, Lifestyle lifestyle)
            => new FactoryRegistration<TTarget>(container, lifestyle);

        public static FactoryRegistration<TTarget> SingletonFactoryFor<TTarget>(this Container container)
            => container.FactoryFor<TTarget>(Lifestyle.Singleton);
    }

...

    container.SingletonFactoryFor<IClickStats>().Register(_ => Settings.Clicks);
    container.RegisterSingleton<IClickStats, Stats>();

    container.SingletonFactoryFor<IImpressionStats>().Register(_ => Settings.Impressions);
    container.RegisterSingleton<IImpressionStats, Stats>();
```

**Comment by dotnetjunkie** [comment]
I can confirm this is a bug. The following test proves the bug:

``` c#
interface IGeneric<T> { }
class IntGenericType : IGeneric<int> { }
class Consumer { public Consumer(IGeneric<int> dependency) { } }

[TestMethod]
public void Test()
{
    // Arrange
    var container = ContainerFactory.New();

    container.RegisterConditional(typeof(IGeneric<int>),
        Lifestyle.Singleton.CreateRegistration<IGeneric<int>>(() => new IntGenericType(), container),
        c => true);

    container.RegisterConditional(typeof(IGeneric<int>),
        Lifestyle.Singleton.CreateRegistration<IGeneric<int>>(() => new IntGenericType(), container),
        c => false);

    // Act
    // This call is suppost to succeed, but it fails.
    container.GetInstance<Consumer>();
}
```

**Comment by dotnetjunkie** [comment]
Fixed in v3.1.3.

**Comment by onyxmaster** [comment]
Thank all of you very much!
